praxis/optimizers.py

```diff
@@ -2775,8 +2777,54 @@ def update_fn(updates: NestedJTensor,
                          lambda: new_count)
 
     def _run_base_tx():
+      
+      def _compute_grad_norm(grads: NestedMap) -> JTensor:
+        """Computes total grad norm."""
+        grad_norms_squared = jax.tree_map(lambda x: jnp.sum(x * x), grads)
+        grad_norms_squared, _ = jax.tree_util.tree_flatten(grad_norms_squared)
+        return jnp.sqrt(jnp.sum(jnp.stack(grad_norms_squared)))
+      
+      
+      def scale_gradients(
+          raw_grads: NestedMap,
+          clip_grad_norm_to_value: Optional[float] = None,
+          clip_grad_single_norm_to_value: Optional[float] = None):
+        
+        def clip_grads(grads, grad_norm):
+          if clip_grad_norm_to_value:
+            assert clip_grad_single_norm_to_value == 0.
+            grad_scale = jnp.minimum(
+                jnp.array(1, grad_norm.dtype),
+                jnp.array(clip_grad_norm_to_value, grad_norm.dtype)
+                / grad_norm)
+            grads = jax.tree_map(lambda g: g * grad_scale, grads)
+          elif clip_grad_single_norm_to_value:
+            assert clip_grad_norm_to_value == 0.
+            grad_single_norm = jax.tree_map(lambda x: jnp.sqrt(jnp.sum(x * x)),
+                                          grads)
+            
+            def scale_gradient(grad, norm):
+              return grad * jnp.minimum(
+                  jnp.array(1, norm.dtype),
+                  jnp.array(clip_grad_single_norm_to_value,
+                            norm.dtype) / norm)
+            grads = jax.tree_map(scale_gradient, grads, grad_single_norm)
+            grad_scale = jnp.array(1.0)
+          else:
+            # no clipping is needed.
+            grad_scale = jnp.array(1.0)
+          return grads, grad_scale
+    
+        raw_grad_norm = _compute_grad_norm(raw_grads)
+        
+        grads, grad_scale = clip_grads(raw_grads, raw_grad_norm)
```

Comment:
do we need to compute and return `grad_scale`?

Reply:
This is not needed. I no longer return `grad_scale` with the latest commit